android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt

```diff
@@ -0,0 +1,164 @@
+        val file = File(path)
+
+        if (file.exists()) {
+            try {
+                val fileReader = FileReader(file)
+                val jsonObject = JSONObject(fileReader.readText())
+                val jsonArray = jsonObject.getJSONArray("records")
+                val records = mutableListOf<RemoteSettingsRecord>()
+
+                for (i in 0 until jsonArray.length()) {
+                    val recordJson = jsonArray.getJSONObject(i)
+                    val id = recordJson.getString("id")
+                    val lastModified = recordJson.getString("lastModified").toULong()
+                    val deleted = recordJson.getBoolean("deleted")
+                    val attachment = recordJson.opt("attachment")
+                    val fields = recordJson.opt("fields")
+
+                    records.add(
+                        RemoteSettingsRecord(
+                            id = id,
+                            lastModified = lastModified,
+                            deleted = deleted,
+                            attachment = attachment as Attachment?,
+                            fields = fields as RsJsonObject
+                        )
+                    )
+                }
+
+                fileReader.close()
+                RemoteSettingsResult.Success(records)
+            } catch (e: Exception) {
+                e.printStackTrace()
+                RemoteSettingsResult.DiskFailure(e)
+            }
+        } else {
+            RemoteSettingsResult.DiskFailure(FileNotFoundException("File not found"))
+        }
+    }
+
+}
+/**
+ * Reads files from local storage. If file is empty, fetches teh file from the remote server.
+ **/
+suspend fun RemoteSettingsClient.readOrFetch(): ReadOrFetchResult {
+
+    return when (val storageResult = read()) {
+        is RemoteSettingsResult.Success -> ReadOrFetchResult.Success(storageResult.records)
+        is RemoteSettingsResult.DiskFailure -> {
+            when (val remoteResult = fetch()) {
+                is RemoteSettingsResult.Success -> ReadOrFetchResult.Success(remoteResult.records)
+                is RemoteSettingsResult.NetworkFailure -> ReadOrFetchResult.Failure(remoteResult.error)
+                is RemoteSettingsResult.DiskFailure -> ReadOrFetchResult.Failure(remoteResult.error)
+
+            }
+        }
+        is RemoteSettingsResult.NetworkFailure -> ReadOrFetchResult.Failure(storageResult.error)
+
+    }
+}
+sealed class RemoteSettingsResult {
+    data class Success(val records: List<RemoteSettingsRecord>) : RemoteSettingsResult()
+    data class DiskFailure(val error: Exception) : RemoteSettingsResult()
+    data class NetworkFailure(val error: Exception) : RemoteSettingsResult()
+}
+
+sealed class ReadOrFetchResult {
+    data class Success(val records: List<RemoteSettingsRecord>) : ReadOrFetchResult()
+    data class Failure(val error: Exception) : ReadOrFetchResult()
+}
```

Comment:
This definitely works, but I think we could get away with just `RemoteSettingsResult`. That way, the return type from the convenience methods and the more focused methods can all be the same.